### ISSUE TITLE
chpasswd.expire = true doesn't work for hashed password

### ISSUE THREAD

**Issue by CrazyKidJack** [canonical/cloud-init#4974]
# Bug report

When this configuration is present in user-data with cloud-init versions > `22.3`, I expect that the specified user (ubuntu) will be asked to reset their password on first password-based login.

```yml
chpasswd:
  expire: true
  users:
    - name: ubuntu
      password: {someHash}
```

I am a software dev, so I tried to look at the source code to figure out what I was doing wrong... but I couldn't figure it out. I also didn't see any obvious bug in [cc_set_passwords.py](https://github.com/canonical/cloud-init/blob/ubuntu/23.3.3-0ubuntu0_22.04.1/cloudinit/config/cc_set_passwords.py). I thought about running the code in a debugger... but thought I'd submit an issue here before getting in that deep.

Note: When trying this setup with a "text" type password, the user is asked to change their password on initial login as expected.
Note 2: I know that in previous version of cloud-init, `expire` only applied to `text` and `RANDOM` passwords... but I'm pretty sure the version I'm using is supposed to support expiration of `hash` passwords as well (at least according to the documentation for [23.3.3](https://cloudinit.readthedocs.io/en/23.3.3/reference/modules.html#set-passwords:~:text=Prior%20to%20cloud%2Dinit%2022.3%2C%20the%20expire%20key%20only%20applies%20to%20plain%20text%20(including%20RANDOM)%20passwords.%20Post%2022.3%2C%20the%20expire%20key%20applies%20to%20both%20plain%20text%20and%20hashed%20passwords.) and [latest](https://cloudinit.readthedocs.io/en/latest/reference/modules.html#set-passwords:~:text=Prior%20to%20cloud%2Dinit%2022.3%2C%20the%20expire%20key%20only%20applies%20to%20plain%20text%20(including%20RANDOM)%20passwords.%20Post%2022.3%2C%20the%20expire%20key%20applies%20to%20both%20plain%20text%20and%20hashed%20passwords.)).

## Steps to reproduce the problem

Base ubuntu minimal cloud image: https://cloud-images.ubuntu.com/minimal/releases/jammy/release-20240207/ubuntu-22.04-minimal-cloudimg-amd64.img

<details>
<summary>List of modifications to that cloud image:</summary>

  - installed vim
  - ran `sudo apt update && sudo apt -y dist-upgrade && sudo apt clean`
  - installed man pages by running a modified version of ubuntu's `unminimize` script (the full, modified script is listed below)

</details>

<details>
<summary>User data:</summary>

  ```yml
  #cloud-config
  hostname: {someName}
  manage_etc_hosts: true
  fqdn: {someName}
  ssh_authorized_keys:
    - ssh-ed25519 {pubKey} {identity}
  chpasswd:
    expire: true
    users:
      - name: ubuntu
        password: $6$rounds=500000$12ko0pUf5xFCtxdt$71TBRBqjmXuJXzUqQgqaMqNVmGcSh91UxNIHdBkeHGbX/RYO5GX53PR2pHmpE/307B7TVIiJLjmwG.kYfygpF0
  ```

</details>

<details>
<summary><code>/etc/cloud/cloud.cfg</code> (default from ubuntu minimal cloud image):</summary>

  ```yml
  # The top level settings are used as module
  # and base configuration.
  
  # A set of users which may be applied and/or used by various modules
  # when a 'default' entry is found it will reference the 'default_user'
  # from the distro configuration specified below
  users:
    - default
  
  # If this is set, 'root' will not be able to ssh in and they
  # will get a message to login instead as the default $user
  disable_root: true
  
  # This will cause the set+update hostname module to not operate (if true)
  preserve_hostname: false
  
  # If you use datasource_list array, keep array items in a single line.
  # If you use multi line array, ds-identify script won't read array items.
  # Example datasource config
  # datasource:
  #   Ec2:
  #     metadata_urls: [ 'blah.com' ]
  #     timeout: 5 # (defaults to 50 seconds)
  #     max_wait: 10 # (defaults to 120 seconds)
  
  # The modules that run in the 'init' stage
  cloud_init_modules:
    - migrator
    - seed_random
    - bootcmd
    - write_files
    - growpart
    - resizefs
    - disk_setup
    - mounts
    - set_hostname
    - update_hostname
    - update_etc_hosts
    - ca_certs
    - rsyslog
    - users_groups
    - ssh
  
  # The modules that run in the 'config' stage
  cloud_config_modules:
    - wireguard
    - snap
    - ubuntu_autoinstall
    - ssh_import_id
    - keyboard
    - locale
    - set_passwords
    - grub_dpkg
    - apt_pipelining
    - apt_configure
    - ubuntu_advantage
    - ntp
    - timezone
    - disable_ec2_metadata
    - runcmd
    - byobu
  
  # The modules that run in the 'final' stage
  cloud_final_modules:
    - package_update_upgrade_install
    - fan
    - landscape
    - lxd
    - ubuntu_drivers
    - write_files_deferred
    - puppet
    - chef
    - ansible
    - mcollective
    - salt_minion
    - reset_rmc
    - rightscale_userdata
    - scripts_vendor
    - scripts_per_once
    - scripts_per_boot
    - scripts_per_instance
    - scripts_user
    - ssh_authkey_fingerprints
    - keys_to_console
    - install_hotplug
    - phone_home
    - final_message
    - power_state_change
  
  # System and/or distro specific settings
  # (not accessible to handlers/transforms)
  system_info:
    # This will affect which distro class gets used
    distro: ubuntu
    # Default user name + that default users groups (if added/used)
    default_user:
      name: ubuntu
      lock_passwd: True
      gecos: Ubuntu
      groups: [adm, audio, cdrom, dialout, dip, floppy, lxd, netdev, plugdev, sudo, video]
      sudo: ["ALL=(ALL) NOPASSWD:ALL"]
      shell: /bin/bash
    network:
      renderers: ['netplan', 'eni', 'sysconfig']
      activators: ['netplan', 'eni', 'network-manager', 'networkd']
    # Automatically discover the best ntp_client
    ntp_client: auto
    # Other config here will be given to the distro class and/or path classes
    paths:
      cloud_dir: /var/lib/cloud/
      templates_dir: /etc/cloud/templates/
    package_mirrors:
      - arches: [i386, amd64]
        failsafe:
          primary: http://archive.ubuntu.com/ubuntu
          security: http://security.ubuntu.com/ubuntu
        search:
          primary:
            - http://%(ec2_region)s.ec2.archive.ubuntu.com/ubuntu/
            - http://%(availability_zone)s.clouds.archive.ubuntu.com/ubuntu/
            - http://%(region)s.clouds.archive.ubuntu.com/ubuntu/
          security: []
      - arches: [arm64, armel, armhf]
        failsafe:
          primary: http://ports.ubuntu.com/ubuntu-ports
          security: http://ports.ubuntu.com/ubuntu-ports
        search:
          primary:
            - http://%(ec2_region)s.ec2.ports.ubuntu.com/ubuntu-ports/
            - http://%(availability_zone)s.clouds.ports.ubuntu.com/ubuntu-ports/
            - http://%(region)s.clouds.ports.ubuntu.com/ubuntu-ports/
          security: []
      - arches: [default]
        failsafe:
          primary: http://ports.ubuntu.com/ubuntu-ports
          security: http://ports.ubuntu.com/ubuntu-ports
    ssh_svcname: ssh
  ```

</details>

<details>
<summary>Modified ubuntu <code>unminimize</code> script:</summary>

  ```sh
  #!/bin/sh
  
  set -e
  
  echo "This system has been minimized by removing packages and content that are"
  echo "not required on a system that users do not log into."
  echo ""
  echo "This script restores content and packages that are found on a default"
  echo "Ubuntu server system in order to make this system more suitable for"
  echo "interactive use."
  echo ""
  echo "Reinstallation of packages may fail due to changes to the system"
  echo "configuration, the presence of third-party packages, or for other"
  echo "reasons."
  echo ""
  echo "This operation may take some time."
  echo ""
  read -p "Would you like to continue? [y/N] " REPLY
  echo    # (optional) move to a new line
  if [ "$REPLY" != "y" ] && [ "$REPLY" != "Y" ]
  then
      exit 1
  fi
  
  if [ -f /etc/dpkg/dpkg.cfg.d/excludes ] || [ -f /etc/dpkg/dpkg.cfg.d/excludes.dpkg-tmp ]; then
      echo "Re-enabling installation of all documentation in dpkg..."
      if [ -f /etc/dpkg/dpkg.cfg.d/excludes ]; then
          mv /etc/dpkg/dpkg.cfg.d/excludes /etc/dpkg/dpkg.cfg.d/excludes.dpkg-tmp
      fi
      echo "Updating package list and upgrading packages..."
      apt-get update
      # apt-get upgrade asks for confirmation before upgrading packages to let the user stop here
      apt-get upgrade
      echo "Restoring system documentation..."
      echo "Reinstalling packages with files in /usr/share/man/ ..."
      # Reinstallation takes place in two steps because a single dpkg --verified
      # command generates very long parameter list for "xargs dpkg -S" and may go
      # over ARG_MAX. Since many packages have man pages the second download
      # handles a much smaller amount of packages.
      dpkg -S /usr/share/man/ |sed 's|, |\n|g;s|: [^:]*$||' | DEBIAN_FRONTEND=noninteractive xargs apt-get install --reinstall -y
      echo "Reinstalling packages with system documentation in /usr/share/doc/ .."
      # This step processes the packages which still have missing documentation
      dpkg --verify --verify-format rpm | awk '$2 ~ /\/usr\/share\/doc/ {print $2}' | sed 's|/[^/]*$||' | sort | uniq \
           | xargs dpkg -S | sed 's|, |\n|g;s|: [^:]*$||' | uniq | DEBIAN_FRONTEND=noninteractive xargs apt-get install --reinstall -y
      #echo "Restoring system translations..."
      # This step processes the packages which still have missing translations
      #dpkg --verify --verify-format rpm | awk '$2 ~ /\/usr\/share\/locale/ {print $2}' | sed 's|/[^/]*$||' | sort | uniq \
      #     | xargs dpkg -S | sed 's|, |\n|g;s|: [^:]*$||' | uniq | DEBIAN_FRONTEND=noninteractive xargs apt-get install --reinstall -y
      if dpkg --verify --verify-format rpm | awk '$2 ~ /\/usr\/share\/doc/ {exit 1}'; then
          echo "Documentation has been restored successfully."
          rm /etc/dpkg/dpkg.cfg.d/excludes.dpkg-tmp
      else
          echo "There are still files missing from /usr/share/doc/:"
          dpkg --verify --verify-format rpm | awk '$2 ~ /\/usr\/share\/doc/ {print " " $2}'
          echo "You may want to try running this script again or you can remove"
          echo "/etc/dpkg/dpkg.cfg.d/excludes.dpkg-tmp and restore the files manually."
      fi
  fi
  
  if  [ "$(dpkg-divert --truename /usr/bin/man)" = "/usr/bin/man.REAL" ]; then
      # Remove diverted man binary
      rm -f /usr/bin/man
      dpkg-divert --quiet --remove --rename /usr/bin/man
  fi
  
  #if ! dpkg-query --show --showformat='${db:Status-Status}\n' ubuntu-minimal 2> /dev/null | grep -q '^installed$'; then
  #    echo "Installing ubuntu-minimal package to provide the familiar Ubuntu minimal system..."
  #    DEBIAN_FRONTEND=noninteractive apt-get install -y ubuntu-minimal ubuntu-standard
  #fi
  
  #if dpkg-query --show --showformat='${db:Status-Status}\n' ubuntu-server 2> /dev/null | grep -q '^installed$' \
  #   && ! dpkg-query --show --showformat='${db:Status-Status}\n' landscape-common 2> /dev/null | grep -q '^installed$'; then
  #    echo "Installing ubuntu-server recommends..."
  #    DEBIAN_FRONTEND=noninteractive apt-get install -y landscape-common
  #fi
  
  # installing LXD using the lxd-installer by simply invoking it
  #echo "Invoking LXD so that it can be installed by the lxd-installer's script.."
  #lxd --version
  
  #echo "Removing lxd installer package..."
  #apt-get purge -y lxd-installer
  
  
  # unminimization succeeded, there is no need to mention it in motd
  rm -f /etc/update-motd.d/60-unminimize
  ```

</details>

<details>
<summary>Custom cloud-init template preparation script (it assumes that it's own name is <code>~/templatize.sh</code> ):</summary>

```bash
#!/bin/bash

rm -rf /etc/ssh/*_host_*
rm -rf /etc/sysctl.d/99-cloudimg-ipv6.conf
rm -rf /etc/default/grub.d/50-cloudimg-settings.cfg
rm -rf /etc/sudoers.d/90-cloud-init-users
rm -rf /etc/ssh/sshd_config.d/60-cloudimg-settings.conf

truncate -s0 /etc/hostname
hostnamectl set-hostname localhost

rm /etc/netplan/50-cloud-init.yaml

truncate -s0 /etc/machine-id
rm /var/lib/dbus/machine-id
ln -s /etc/machine-id /var/lib/dbus/machine-id

cloud-init clean

passwd -dl root

truncate -s0 ~/.bash_history
history -c

chmod 400 ~/templatize.sh

shutdown -h now
```

</details>

<details open>
<summary><b>Steps to reproduce:</b></summary>

1. Download the ubuntu minimal cloud image
2. Boot the cloud image and make listed modifications (this may not actually be necessary for reproduction)
3. Prepare the image to be used as a custom cloud-init template following this guide: https://medium.com/@dsykes012/making-a-custom-ubuntu-20-04-lts-focal-fossa-vm-template-that-works-with-cloud-init-2cfffb6783b4 (if step 2 is skipped, this may also be skipped)
  a. I provided a script that will do the preparation for you.
5. Create the specified user-data file
6. Boot the image again, providing the associated user-data (I did it using proxmox VE)
7. Login with the default user (`ubuntu`) using the password. (The password specified by the hash in the provided user-data file is "`changeme`")
8. Observe that the user is NOT asked to change their password

</details>

## Environment details
- Cloud-init version:
```bash
$ cloud-init --version
/usr/bin/cloud-init 23.4.3-0ubuntu0~22.04.1
```
(I also tried with 23.3.3-0ubuntu0~22.04.1)

- Operating System Distribution:
```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.4 LTS
Release:        22.04
Codename:       jammy

$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

- Cloud provider, platform or installer type:
  - Downloaded this cloud image: https://cloud-images.ubuntu.com/minimal/releases/jammy/release-20240207/ubuntu-22.04-minimal-cloudimg-amd64.img
  - Deployed on proxmox VE 8.1.4

## cloud-init logs

- [full redacted cloud-init.log](https://github.com/canonical/cloud-init/files/14429889/cloud-init.log)
- [full redacted cloud-init-output.log](https://github.com/canonical/cloud-init/files/14429892/cloud-init-output.log)
- [My best guess at the relevant part of cloud-init.log](https://github.com/canonical/cloud-init/files/14429908/cloud-init.log.excerpt.log)... though I don't think it will be very useful :/ Just shows that it did not try to expire the password

**Comment by catmsred** [comment]
Thank you for reporting this issue!  While expiring hashed passwords is available in cloud-init 22.3 and beyond, it is patched out on ubuntu releases 22.04 and earlier to ensure consistent behaviour for the lifetime of those releases.  If you use a newer version of ubuntu, you will see expire respected for hashed users as well.

**Comment by CrazyKidJack** [comment]
@catmsred  Thanks for the speedy response! Is there someone where that information was posted so I could have known it without bothering you with an issue? That way I know where to look next time I have a problem
